tests/test_raster_index.py

```
@@ -31,3 +34,195 @@ def test_sel_slice():
+            [
+                # increasing Δy
+                "-40.0 5 0.0 -2.0 0.0 0.5",
+                "-40.0 5 0.0 0.0 0.0 0.5",
+                "-40.0 5 0.0 2.0 0.0 0.5",
+            ],
+            "y",
+        ),
+    ],
+)
+def test_concat_and_combine_nested_1D(transforms, concat_dim):
+    """Models two side-by-side tiles"""
+    crs_attrs = pyproj.CRS.from_epsg(4326).to_cf()
+
+    datasets = [
+        xr.Dataset(
+            {"foo": (("y", "x"), np.ones((4, 2)), {"grid_mapping": "spatial_ref"})},
+            coords={"spatial_ref": ((), 0, crs_attrs | {"GeoTransform": transform})},
+        )
+        for transform in transforms
+    ]
+    datasets = list(map(assign_index, datasets))
+
+    if concat_dim == "x":
+        new_data = np.ones((4, 2 * len(transforms)))
+    else:
+        new_data = np.ones((4 * len(transforms), 2))
+    expected = xr.Dataset(
+        {"foo": (("y", "x"), new_data, {"grid_mapping": "spatial_ref"})},
+        coords={"spatial_ref": ((), 0, crs_attrs | {"GeoTransform": transforms[0]})},
+    ).pipe(assign_index)
+
+    for actual in [
+        xr.combine_nested(datasets, concat_dim=concat_dim, combine_attrs="override"),
+        xr.concat(datasets, dim=concat_dim),
+    ]:
+        assert_identical(actual, expected)
+        assert_identical(actual, expected)
+        concat_coord = xr.concat([ds[concat_dim] for ds in datasets], dim=concat_dim)
+        assert_identical(actual[concat_dim], concat_coord)
+
+
+@pytest.mark.parametrize(
+    "transforms, concat_dim",
+    [
+        (
+            [
+                # out-of-order for Y
+                "-40.0 5 0.0 -2.0 0.0 -0.5",
+                "-40.0 5 0.0 0.0 0.0 -0.5",
+                "-40.0 5 0.0 2.0 0.0 -0.5",
+            ],
+            "y",
+        ),
+        (
+            [
+                # incompatible, different origins
+                "-50.0 5 0.0 -2.0 0.0 -0.5",
+                "-40.0 5 0.0 0.0 0.0 -0.5",
+            ],
+            "x",
```

@benbovy this one is triggering an AssertionError inside Xarray: https://github.com/pydata/xarray/blob/d8a4a05df67681407e05c3a1c04d79e46deeaffe/xarray/structure/alignment.py#L577

This seems problematic but I don't fully understand it. Is the indent on these two lines wrong? 
https://github.com/pydata/xarray/blob/d8a4a05df67681407e05c3a1c04d79e46deeaffe/xarray/structure/alignment.py#L277-L278

---

The indent seems right to me.

When excluding dimension "x" from the alignment like in this case, we actually don't want to assign a new, common raster index (computed from `RasterIndex.join()`) to each dataset in the aligned results. Instead we want to preserve the original indexes of each dataset along "x".

The `assert self.reindex[((("x", ("x",)), ("y", ("y"))), RasterIndex)] is False` in Xarray ensures that. If the assertion passes, it means that all raster indexes are equal along "y" (regardless of the `join` option) so we can preserve the original raster indexes in the aligned datasets. This test shows it may not always be the case (datasets are not aligned along "y") so we'll need to raise a better error than the AssertionError + an informative error message.

---

In theory it might be possible to support `xr.align(*datasets, exclude="x", join="outer")` here, although this would require to somehow pass `exclude="x"` to `RasterIndex.join()` and compute / assign a different aligned index for each dataset such that it preserves the original labels along "x". This would make the Xarray alignment logic much more complicated...